python/tchannel/sync/client.py

```diff
@@ -0,0 +1,130 @@
+
+
+class TChannelClient(object):
+    """Implements a synchronous TChannel client.
+
+    The client does not support receiving incoming connections and therefore,
+    is not really bidirectional.
+
+    The client is implemented on top of the Tornado-based implementation and
+    starts and stops IOLoops on-demand.
+
+    .. code-block:: python
+
+        client = TChannelClient()
+        response = client.request(
+            hostport='localhost:4040',
+            service='HelloService',
+        ).send(
+            'hello', None, json.dumps({"name": "World"})
+        )
+
+    """
+
+    def __init__(self, process_name=None, io_loop=None):
+        """Initialize a new TChannelClient.
+
+        :param process_name:
+            Name of the calling process. Used for logging purposes only.
+        :param io_loop:
+            Tornado IOLoop through which requests will be made.
+        """
+        self.io_loop = ioloop.IOLoop.current()
+        self.async_client = async.TChannel(
+            hostport=glossary.EPHEMERAL_HOSTPORT,
+            process_name=process_name,
+        )
+
+    def request(self, *args, **kwargs):
+        """Initiate a new request to a peer.
+
+        :param hostport:
+            If specified, requests will be sent to the specific host.
+            Otherwise, a known peer will be picked at random.
+        :param service:
+            Name of the service being called. Defaults to an empty string.
+        :param service_threshold:
+            If ``hostport`` was not specified, this specifies the score
+            threshold at or below which peers will be ignored.
+        :returns:
+            An object with a ``send(arg1, arg2, arg3)`` operation.
+        """
+        return ClientOperationWrapper(
+            self.io_loop, self.async_client.request(*args, **kwargs)
+        )
+
+
+class ClientOperationWrapper(object):
+    """Allows making client operation requests synchronously."""
+
+    def __init__(self, io_loop, ops):
+        assert io_loop, "io_loop is required"
+        assert ops, "ops is required"
+
+        self.io_loop = io_loop
+        self.ops = ops
+
+    def send(self, arg1, arg2, arg3):
+        """Send the given triple over the wire.
+
+        :param arg1:
+            String containing the contents of arg1. If None, an empty string
+            is used.
+        :param arg2:
+            String containing the contents of arg2. If None, an empty string
+            is used.
+        :param arg3:
+            String containing the contents of arg3. If None, an empty string
+            is used.
+        :return:
+            Response from the peer.
+        """
+        arg1 = arg1 or ''
+        arg2 = arg2 or ''
+        arg3 = arg3 or ''
+
+        @gen.coroutine
+        def go():
+            # This is the response that contains argstreams. We need to read
+            # everything in memory.
+            resp = yield self.ops.send(arg1, arg2, arg3)
+
```

Comment:
It will throw an exception if there's an error. The future will then contain an exception which will be re-raised by the run_sync call.